File: java/src/main/java/net/imagini/aim/loaders/Loader.java
package net.imagini.aim.loaders;

import java.io.EOFException;
import java.io.IOException;
import java.net.InetAddress;
import java.net.Socket;

import net.imagini.aim.AimSchema;
import net.imagini.aim.AimType;
import net.imagini.aim.Pipe;
import net.imagini.aim.PipeLZ4;

public abstract class Loader extends Thread {
    private Socket socket;
    private Pipe out;
    protected AimSchema schema;

    protected void connectTable(String host, int port, AimSchema schema) throws IOException {
        socket = new Socket(InetAddress.getByName(host), port);
        out = new PipeLZ4(socket.getOutputStream(), Pipe.Protocol.LOADER);
        out.write(schema.toString());
        this.schema = schema;
    }
    protected void close() throws IOException {
        out.close();
        socket.close();
    }

    protected byte[][] createEmptyRecord() {
        return new byte[schema.size()][];
    }

    protected void storeLoadedRecord(byte[][] record) throws IOException {
        try {
            int i = 0;
            for (AimType type : schema.def()) {
                out.write(type.getDataType(), record[i]);
                i++;
            }
        } catch (EOFException e) {
            out.flush();
            throw e;
        }
    }
}

File: scala/src/main/scala/net/imagini/aim/cluster/Loader.scala
package net.imagini.aim.cluster

import java.io.BufferedReader
import java.io.EOFException
import java.io.FileInputStream
import java.io.InputStream
import java.io.InputStreamReader
import java.net.InetAddress
import java.net.Socket
import java.util.zip.GZIPInputStream
import scala.Array.canBuildFrom
import grizzled.slf4j.Logger
import net.imagini.aim.types.AimSchema
import net.imagini.aim.types.AimType
import java.io.IOException
import net.imagini.aim.tools.StreamUtils
import net.imagini.aim.types.Aim

object Loader extends App {
  var host: String = "localhost"
  var port: Int = 4000
  var keyspace: String = null
  var table: String = null
  var separator: String = "\n"
  var gzip = false
  var file: Option[String] = None
  val argsIterator = args.iterator
  while (argsIterator.hasNext) {
    argsIterator.next match {
      case "--host"      ⇒ host = argsIterator.next
      case "--port"      ⇒ port = argsIterator.next.toInt
      case "--keyspace"  ⇒ keyspace = argsIterator.next
      case "--table"     ⇒ table = argsIterator.next
      case "--separator" ⇒ separator = argsIterator.next
      case "--file"      ⇒ file = Some(argsIterator.next)
      case "--gzip"      ⇒ gzip = true
      case arg: String   ⇒ println("Unknown argument " + arg)
    }
  }
  val loader = file match {
    case None           ⇒ new Loader(host, port, keyspace, table, separator, gzip)
    case Some(filename) ⇒ new Loader(host, port, keyspace, table, separator, filename, gzip)
  }
  loader.streamInput
}

class Loader(host: String, port: Int, val keyspace: String, val table: String, val separator: String, val fileinput: InputStream, val gzip: Boolean) {

  def this(host: String, port: Int, keyspace: String, table: String, separator: String, filename: String, gzip: Boolean) = this(host, port, keyspace, table, separator, if (filename == null) null else new FileInputStream(filename), gzip)

  def this(host: String, port: Int, keyspace: String, table: String, separator: String, gzip: Boolean) = this(host, port, keyspace, table, separator, null.asInstanceOf[InputStream], gzip)

  val log = Logger[this.type]
  val in: InputStream = if (fileinput == null) System.in else fileinput
  val socket = new Socket(InetAddress.getByName(host), port)
  val pipe = new PipeGZIPLoader(socket, Protocol.LOADER_LOCAL)

  //handshake
  pipe.writeDirect(keyspace)
  pipe.writeDirect(table)
  pipe.writeDirect(separator)
  if (gzip) {
    pipe.getDirectOutputStream.finish
  } else {
    pipe.getDirectOutputStream.flush
  }
  val schemaDeclaration = pipe.read
  val schema = AimSchema.fromString(schemaDeclaration)

  def streamInput:Int = {
    org.apache.commons.io.IOUtils.copy(in, if (gzip) pipe.getOutputStream() else pipe.getDirectOutputStream())
    if (!gzip) pipe.getDirectOutputStream.finish
    pipe.getOutputStream().flush
    val loadedCount:Int = pipe.readInt
    pipe.close
    socket.close
    loadedCount 
  }

}